x-pack/plugins/alerting/public/pages/maintenance_windows/components/table_actions_popover.tsx

```diff
@@ -0,0 +1,209 @@
+  EuiContextMenuPanel,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiPopover,
+} from '@elastic/eui';
+import * as i18n from '../translations';
+import { MaintenanceWindowStatus } from '../../../../common';
+
+interface TableActionsPopoverProps {
+  status: MaintenanceWindowStatus;
+  onEdit: () => void;
+  onCancel: () => void;
+  onArchive: (archive: boolean) => void;
+  onCancelAndArchive: () => void;
+}
+type ModalType = 'cancel' | 'cancelAndArchive' | 'archive' | 'unarchive';
+
+export const TableActionsPopover: React.FC<TableActionsPopoverProps> = React.memo(
+  ({ status, onEdit, onCancel, onArchive, onCancelAndArchive }) => {
+    const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+    const [isModalVisible, setIsModalVisible] = useState(false);
+    const [modalType, setModalType] = useState<ModalType>();
+
+    const onButtonClick = useCallback(() => {
+      setIsPopoverOpen((open) => !open);
+    }, []);
+    const closePopover = useCallback(() => {
+      setIsPopoverOpen(false);
+    }, []);
+
+    const closeModal = useCallback(() => setIsModalVisible(false), []);
+    const showModal = useCallback((type: ModalType) => {
+      setModalType(type);
+      setIsModalVisible(true);
+    }, []);
+
+    const modal = useMemo(() => {
+      const modals = {
+        cancel: {
+          title: i18n.CANCEL_MODAL_TITLE,
+          subtitle: i18n.CANCEL_MODAL_SUBTITLE,
+          cancelButton: i18n.CANCEL_MODAL_BUTTON,
+          onConfirm: () => onCancel(),
+        },
+        cancelAndArchive: {
+          title: i18n.CANCEL_AND_ARCHIVE_MODAL_TITLE,
+          subtitle: i18n.CANCEL_AND_ARCHIVE_MODAL_SUBTITLE,
+          cancelButton: i18n.CANCEL_MODAL_BUTTON,
+          onConfirm: () => onCancelAndArchive(),
+        },
+        archive: {
+          title: i18n.ARCHIVE_TITLE,
+          subtitle: i18n.ARCHIVE_SUBTITLE,
+          cancelButton: i18n.CANCEL,
+          onConfirm: () => onArchive(true),
+        },
+        unarchive: {
+          title: i18n.UNARCHIVE_MODAL_TITLE,
+          subtitle: i18n.UNARCHIVE_MODAL_SUBTITLE,
+          cancelButton: i18n.CANCEL,
+          onConfirm: () => onArchive(false),
+        },
+      };
+      let m;
+      if (isModalVisible && modalType) {
+        const modalProps = modals[modalType];
+        m = (
+          <EuiConfirmModal
+            style={{ width: 600 }}
+            title={modalProps.title}
+            onCancel={closeModal}
+            onConfirm={() => {
+              closeModal();
+              modalProps.onConfirm();
+            }}
+            cancelButtonText={modalProps.cancelButton}
+            confirmButtonText={modalProps.title}
+            defaultFocusedButton="confirm"
+            buttonColor="danger"
+          >
+            <p>{modalProps.subtitle}</p>
+          </EuiConfirmModal>
+        );
+      }
+      return m;
+    }, [modalType, isModalVisible, closeModal, onArchive, onCancel, onCancelAndArchive]);
+
+    const items = useMemo(() => {
```

Comment:
I like what you did up there with the `modals` map, but I'm wondering if we should do the same for the menu items since it's a bit hard to untangle what status gets what menu items:

```tsx
const menuItemMap = {
  edit: <EuiContextMenuItem {...props} />,
  cancel: <EuiContextMenuItem {...props} />,
  cancelAndArchive: <EuiContextMenuItem {...props} />,
  archive: <EuiContextMenuItem {...props} />,
  unarchive: <EuiContextMenuItem {...props} />,
}

const statusMenuItemsTypesMap = {
  archived: ['cancel', ...],
  running: [...],
  ...
}

return statusMenuItemsTypesMap[status].map((type) => menuItemMap[type]));
```

or something like that, play around with the variable naming too, I just thought of some stuff on the spot haha.

Reply:
Resolved in this commit https://github.com/elastic/kibana/pull/155201/commits/9d6e6730e18bd54c5bece8e97a0c801d232edd3c